core/src/main/scala/akka/kafka/internal/SubSourceLogic.scala

```diff
@@ -95,6 +96,15 @@ private abstract class SubSourceLogic[K, V, Msg](
             _.tell(TopicPartitionsRevoked(subscription, revokedTps), sourceActor.ref)
           }
           if (revokedTps.nonEmpty) {
+            if (waitForStreamCompletion) {
+              implicit val ec: ExecutionContext = materializer.executionContext
+              log.debug(s"Shutting down subsources for: $revokedTps")
+              val futures = revokedTps
+                .flatMap(subSources.get)
+                .map { case (control, streamFuture) => control.shutdown().flatMap(_ => streamFuture) }
```

Comment:
I wonder if it needs to handle sub-source failures here. (For a horrible example look at `akka.kafka.scaladsl.Consumer.Control.drainAndShutdown`.)

Reply:
I'll create a test for this.